homeassistant/components/tasmota/device_trigger.py

```diff
@@ -0,0 +1,287 @@
+            """Remove trigger."""
+            if instance not in self.trigger_instances:
+                raise HomeAssistantError("Can't remove trigger twice")
+
+            if instance.remove:
+                instance.remove()
+            self.trigger_instances.remove(instance)
+
+        return async_remove
+
+    def detach_trigger(self):
+        """Remove Tasmota device trigger."""
+        # Mark trigger as unknown
+        self.tasmota_trigger = None
+
+        # Unsubscribe if this trigger is in use
+        for trig in self.trigger_instances:
+            if trig.remove:
+                trig.remove()
+                trig.remove = None
+
+    async def arm_tasmota_trigger(self):
+        """Arm Tasmota trigger: subscribe to MQTT topics and fire events."""
+
+        @callback
+        def _on_trigger():
+            data = {
+                "mac": self.tasmota_trigger.cfg.mac,
+                "source": self.tasmota_trigger.cfg.source,
+                "event": self.tasmota_trigger.cfg.event,
+            }
+            self.hass.bus.async_fire(
+                TASMOTA_EVENT,
+                data,
+            )
+
+        self.tasmota_trigger.set_on_trigger_callback(_on_trigger)
+        await self.tasmota_trigger.subscribe_topics()
+
+    async def set_tasmota_trigger(self, tasmota_trigger, remove_update_signal):
+        """Set Tasmota trigger."""
+        await self.update_tasmota_trigger(tasmota_trigger.cfg, remove_update_signal)
+        self.tasmota_trigger = tasmota_trigger
+
+        for trig in self.trigger_instances:
+            await trig.async_attach_trigger()
+
+    async def update_tasmota_trigger(self, tasmota_trigger_cfg, remove_update_signal):
+        """Update Tasmota trigger."""
+        self.remove_update_signal = remove_update_signal
+        self.type = tasmota_trigger_cfg.type
+        self.subtype = tasmota_trigger_cfg.subtype
+
+
+async def async_setup_trigger(hass, tasmota_trigger, config_entry, discovery_hash):
+    """Set up a discovered Tasmota device trigger."""
+    discovery_id = tasmota_trigger.cfg.trigger_id
+    remove_update_signal = None
+    _LOGGER.debug(
+        "Discovered trigger with id: %s '%s'", discovery_id, tasmota_trigger.cfg
+    )
+
+    async def discovery_update(trigger_config):
+        """Handle discovery update."""
+        _LOGGER.debug(
+            "Got update for trigger with hash: %s '%s'", discovery_hash, trigger_config
+        )
+        if not trigger_config.is_active:
+            # Empty trigger_config: Remove trigger
+            _LOGGER.debug("Removing trigger: %s", discovery_hash)
+            if discovery_id in hass.data[DEVICE_TRIGGERS]:
+                device_trigger = hass.data[DEVICE_TRIGGERS][discovery_id]
+                await device_trigger.tasmota_trigger.unsubscribe_topics()
+                device_trigger.detach_trigger()
+                clear_discovery_hash(hass, discovery_hash)
+                remove_update_signal()
+        else:
```

Comment:
Add returns at the end of the if-blocks instead of else-statements.

Reply:
Thanks, fixed.